Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp

```diff
@@ -2470,20 +2472,19 @@ void CanvasRenderingContext2DBase::putImageData(ImageData& data, int dx, int dy,
         dirtyHeight = -dirtyHeight;
     }
 
-    IntSize destOffset { dx, dy };
+    IntPoint destOffset { dx, dy };
     IntRect destRect { dirtyX, dirtyY, dirtyWidth, dirtyHeight };
-    // FIXME: computeImageDataRect also updates destRect. Maybe return a tuple? Or move
-    // the calculation of the real destRect to here.
-    IntRect sourceRect = computeImageDataRect(*buffer, data.width(), data.height(), destRect, destOffset);
-
-    if (!sourceRect.isEmpty())
-        buffer->putPixelBuffer(data.pixelBuffer(), sourceRect, IntPoint { destOffset });
+    IntRect sourceRect = computeImageDataRect(*buffer, data.size(), destRect, destOffset);
 
     OptionSet<DidDrawOption> options; // ignore transform, shadow, clip, and post-processing
-
-    if (cacheImageDataIfPossible(data, { dx, dy }, sourceRect))
-        options.add(DidDrawOption::PreserveCachedImageData);
-
+    if (!sourceRect.isEmpty()) {
```

Comment:
Can we return early if this is empty? Or is there a need to call `didDraw` for some reason.

Reply:
I'll do that separately